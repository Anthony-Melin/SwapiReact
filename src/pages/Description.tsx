import { LoaderFunctionArgs, useLoaderData } from 'react-router'
import getFetch from '@api/getFetch.ts'
import Details from '@api/types/details.ts'
import LabelValue from '../components/LabelValue.tsx'
import LabelDate from '../components/LabelDate.tsx'
import ListValue from '../components/ListValue.tsx'

export const clientLoader = async ({ request }: LoaderFunctionArgs) => {
    const url = new URL(request.url)
    return {
        details: await getFetch<Details>(url.pathname, request.signal),
    }
}

const Description = () => {
    const {
        details: { name, title, ...restKeys },
    } = useLoaderData() as Awaited<ReturnType<typeof clientLoader>>
    return (
        <main id="description-page" data-testid="description-page">
            <h1>{name || title}</h1>
            {Object.entries(restKeys).map(([key, value]) => {
                if (typeof value === 'string') {
                    if (key === 'height') {
                        return (
                            <LabelValue
                                key={key}
                                label="Height"
                                value={value}
                                testid="height"
                            />
                        )
                    } else if (key === 'mass') {
                        return (
                            <LabelValue
                                key={key}
                                label="Mass"
                                value={value}
                                testid="mass"
                            />
                        )
                    } else if (key === 'hair_color') {
                        return (
                            <LabelValue
                                key={key}
                                label="Hair color"
                                value={value}
                                testid="haircolor"
                            />
                        )
                    } else if (key === 'skin_color') {
                        return (
                            <LabelValue
                                key={key}
                                label="Skin color"
                                value={value}
                                testid="skincolor"
                            />
                        )
                    } else if (key === 'eye_color') {
                        return (
                            <LabelValue
                                key={key}
                                label="Eye color"
                                value={value}
                                testid="eyecolor"
                            />
                        )
                    } else if (key === 'birth_year') {
                        return (
                            <LabelValue
                                key={key}
                                label="Birth year"
                                value={value}
                                testid="birthyear"
                            />
                        )
                    } else if (key === 'gender') {
                        return (
                            <LabelValue
                                key={key}
                                label="Gender"
                                value={value}
                                testid="gender"
                            />
                        )
                    } else if (key === 'homeworld') {
                        return (
                            <LabelValue
                                key={key}
                                label="Home world"
                                value={value}
                                testid="homeworld"
                            />
                        )
                    } else if (key === 'created') {
                        return (
                            <LabelDate
                                key={key}
                                label="Created"
                                value={value}
                                testid="created"
                            />
                        )
                    } else if (key === 'edited') {
                        return (
                            <LabelDate
                                key={key}
                                label="Edited"
                                value={value}
                                testid="edited"
                            />
                        )
                    } else if (key === 'release_date') {
                        return (
                            <LabelValue
                                key={key}
                                label="Release date"
                                value={value}
                                testid="releasedate"
                            />
                        )
                    } else if (key === 'producer') {
                        return (
                            <LabelValue
                                key={key}
                                label="Producer"
                                value={value}
                                testid="producer"
                            />
                        )
                    } else if (key === 'director') {
                        return (
                            <LabelValue
                                key={key}
                                label="Director"
                                value={value}
                                testid="director"
                            />
                        )
                    }
                } else if (typeof value === 'number') {
                    return (
                        <LabelValue
                            key={key}
                            label="Episode"
                            value={value.toString()}
                            testid="episode"
                        />
                    )
                } else if (value instanceof Array) {
                    const values = value.filter((v) => typeof v === 'string')
                    if (key === 'films') {
                        return (
                            <ListValue
                                key={key}
                                title="Films"
                                item="film"
                                values={values}
                                testid="film"
                            />
                        )
                    } else if (key === 'species') {
                        return (
                            <ListValue
                                key={key}
                                title="Species"
                                item="species"
                                values={values}
                                testid="species"
                            />
                        )
                    } else if (key === 'vehicles') {
                        return (
                            <ListValue
                                key={key}
                                title="Vehicles"
                                item="vehicle"
                                values={values}
                                testid="vehicle"
                            />
                        )
                    } else if (key === 'starships') {
                        return (
                            <ListValue
                                key={key}
                                title="Starships"
                                item="starship"
                                values={values}
                                testid="starship"
                            />
                        )
                    } else if (key === 'characters') {
                        return (
                            <ListValue
                                key={key}
                                title="Characters"
                                item="character"
                                values={values}
                                testid="character"
                            />
                        )
                    }
                }
            })}
        </main>
    )
}

export default Description
