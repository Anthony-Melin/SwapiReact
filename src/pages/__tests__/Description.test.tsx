import { beforeEach, describe, expect, it } from 'vitest'
import { render, screen } from '@testing-library/react'
import { createRoutesStub } from 'react-router'
import Description, { clientLoader } from '@pages/Description.tsx'

describe('Description', () => {
    const Stub = createRoutesStub([
        {
            path: "/:section/:id",
            Component: Description,
            loader: clientLoader,
            HydrateFallback: () => <div data-testid="fallback" />,
            ErrorBoundary: () => <div data-testid="error" />
        },
    ])

    const renderDescription = (sufixUrl: string) => {
        render(<Stub initialEntries={[sufixUrl]} />)
    }

    describe('people 1', () => {
        beforeEach(() => {
            renderDescription('/people/1')
        })

        it('should match snapshot', async () => {
            expect(await screen.findByTestId('description-page')).toMatchSnapshot()
        })

        it('should render properties', async () => {
            expect(await screen.findByTestId('height')).toHaveTextContent('Height: 172')

            expect(await screen.findByTestId('mass')).toHaveTextContent('Mass: 77')

            expect(await screen.findByTestId('haircolor')).toHaveTextContent('Hair color: blond')

            expect(await screen.findByTestId('skincolor')).toHaveTextContent('Skin color: fair')

            expect(await screen.findByTestId('eyecolor')).toHaveTextContent('Eye color: blue')

            expect(await screen.findByTestId('birthyear')).toHaveTextContent('Birth year: 19BBY')

            expect(await screen.findByTestId('gender')).toHaveTextContent('Gender: male')

            expect(await screen.findByTestId('homeworld')).toHaveTextContent('Home world: https://swapi.dev/api/planets/1/')

            expect(await screen.findByTestId('film-title')).toHaveTextContent('Films')
            expect(await screen.findByTestId('film-0')).toHaveTextContent('https://swapi.dev/api/films/1/')
            expect(await screen.findByTestId('film-1')).toHaveTextContent('https://swapi.dev/api/films/2/')
            expect(await screen.findByTestId('film-2')).toHaveTextContent('https://swapi.dev/api/films/3/')
            expect(await screen.findByTestId('film-3')).toHaveTextContent('https://swapi.dev/api/films/6/')

            expect(await screen.findByTestId('species-title')).toHaveTextContent('Species')
            expect(await screen.findByTestId('species-empty')).toHaveTextContent('No species linked')

            expect(await screen.findByTestId('vehicle-title')).toHaveTextContent('Vehicles')
            expect(await screen.findByTestId('vehicle-0')).toHaveTextContent('https://swapi.dev/api/vehicles/14/')
            expect(await screen.findByTestId('vehicle-1')).toHaveTextContent('https://swapi.dev/api/vehicles/30/')

            expect(await screen.findByTestId('starship-title')).toHaveTextContent('Starships')
            expect(await screen.findByTestId('starship-0')).toHaveTextContent('https://swapi.dev/api/starships/12/')
            expect(await screen.findByTestId('starship-1')).toHaveTextContent('https://swapi.dev/api/starships/22/')

            expect(await screen.findByTestId('created')).toHaveTextContent('Created: 09/12/2014 14:50:51')

            expect(await screen.findByTestId('edited')).toHaveTextContent('Edited: 20/12/2014 22:17:56')

        })
    })
})
