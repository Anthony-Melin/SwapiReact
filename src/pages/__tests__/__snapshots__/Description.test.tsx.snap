// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Description > people 1 > should match snapshot 1`] = `
<main
  data-testid="description-page"
  id="description-page"
>
  <h1>
    Luke Skywalker
  </h1>
  <p
    data-testid="height"
  >
    Height
    : 
    172
  </p>
  <p
    data-testid="mass"
  >
    Mass
    : 
    77
  </p>
  <p
    data-testid="haircolor"
  >
    Hair color
    : 
    blond
  </p>
  <p
    data-testid="skincolor"
  >
    Skin color
    : 
    fair
  </p>
  <p
    data-testid="eyecolor"
  >
    Eye color
    : 
    blue
  </p>
  <p
    data-testid="birthyear"
  >
    Birth year
    : 
    19BBY
  </p>
  <p
    data-testid="gender"
  >
    Gender
    : 
    male
  </p>
  <p
    data-testid="homeworld"
  >
    Home world
    : 
    https://swapi.dev/api/planets/1/
  </p>
  <div
    data-testid="film"
  >
    <p
      data-testid="film-title"
    >
      Films
    </p>
    <ul>
      <li
        data-testid="film-0"
      >
        https://swapi.dev/api/films/1/
      </li>
      <li
        data-testid="film-1"
      >
        https://swapi.dev/api/films/2/
      </li>
      <li
        data-testid="film-2"
      >
        https://swapi.dev/api/films/3/
      </li>
      <li
        data-testid="film-3"
      >
        https://swapi.dev/api/films/6/
      </li>
    </ul>
  </div>
  <div
    data-testid="species"
  >
    <p
      data-testid="species-title"
    >
      Species
    </p>
    <p
      data-testid="species-empty"
    >
      No 
      species
       linked
    </p>
  </div>
  <div
    data-testid="vehicle"
  >
    <p
      data-testid="vehicle-title"
    >
      Vehicles
    </p>
    <ul>
      <li
        data-testid="vehicle-0"
      >
        https://swapi.dev/api/vehicles/14/
      </li>
      <li
        data-testid="vehicle-1"
      >
        https://swapi.dev/api/vehicles/30/
      </li>
    </ul>
  </div>
  <div
    data-testid="species"
  >
    <p
      data-testid="species-title"
    >
      Starships
    </p>
    <ul>
      <li
        data-testid="species-0"
      >
        https://swapi.dev/api/starships/12/
      </li>
      <li
        data-testid="species-1"
      >
        https://swapi.dev/api/starships/22/
      </li>
    </ul>
  </div>
  <p
    data-testid="created"
  >
    Created
    : 
    09/12/2014 14:50:51
  </p>
  <p
    data-testid="edited"
  >
    Edited
    : 
    20/12/2014 22:17:56
  </p>
</main>
`;
